File: C++/unique_binary_search_tree_II.cpp
//no. 95
#include <iostream>
#include <vector>
#include "tree_node.h"
using namespace std;

class Solution {
    public:
        vector<TreeNode*> generateTrees(int n) {
            return getTrees(1, n);            
        }
        vector <TreeNode*> getTrees(int start, int end){
            vector<TreeNode*> result;
            if (start > end){
                result.push_back(NULL);
                return result;
            }else if(start == end){
                result.push_back(new TreeNode(start));
                return result;
            }
            for(int root=start; root <= end; root ++){
                vector<TreeNode*> lefts = getTrees(start, root-1);
                vector<TreeNode*> rights = getTrees(root+1, end);
                for(int i = 0; i < lefts.size(); i++){
                    for(int j = 0; j < rights.size(); j++){
                        result.push_back(new TreeNode(root, lefts[i], rights[j]));
                    }
                }
            }
            return result;
        }

};

int main(){
    Solution sol;
    for (int i = 0; i <= 8; i++){
        vector<TreeNode*> ans = sol.generateTrees(i);
        cout << i << " " << ans.size() << endl;
        cout << "--------" << endl;
    }
}

File: python/unique_binary_search_tree_II.py
# no.95
from typing import List, Optional
from utils import TreeNode

class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        def get_trees(start, end):
            if start > end:
                return [None]
            elif start == end:
                return [TreeNode(start)]
            result = []
            for root in range(start, end + 1):
                lefts = get_trees(start, root - 1)
                rights = get_trees(root + 1, end)
                for left_tree in lefts:
                    for right_tree in rights:
                        result.append(TreeNode(root, left_tree, right_tree))
            return result
        return get_trees(1, n)
   
if __name__ == '__main__':
    sol = Solution()
    for i in range(1, 9):
        ans = sol.generateTrees(i)
#        print(ans)
        print(len(ans))
        print('-' * 20)
